src/runtime_src/xdp/profile/device/profile_ip_access.cpp

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright (C) 2019-2020, Xilinx Inc - All rights reserved
+ * Xilinx Debug & Profile (XDP) APIs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"). You may
+ * not use this file except in compliance with the License. A copy of the
+ * License is located at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include "profile_ip_access.h"
+
+namespace xdp {
+
+ProfileIP::ProfileIP(xclDeviceHandle handle, int index) {
+    // initialize member variables
+    exclusive = false;
+    mapped = false;
+    mapped_address = 0;
+    device_handle = nullptr;
+    ip_index = -1;
+
+    // check for exclusive access to this IP
+    request_exclusive_ip_access(handle, index);
+    if (exclusive) {
+        device_handle = handle;
+        ip_index = index;
+    } else {
+        show_warning("Cannot get exclusive access");
+    }
+}
+
+ProfileIP::~ProfileIP() {
+    if (mapped) {
+        unmap();
+    }
+    if (exclusive) {
+        release_exclusive_ip_access(device_handle, ip_index);
+    }
+} 
+
+void ProfileIP::request_exclusive_ip_access(xclDeviceHandle handle, int index) {
+    /**
+     * TODO: when the XRT implements the exclusive context hal API, this
+     * method should try to open a exclusive context here and set the
+     * exclusive flag to be true if successful and save the context object
+     * into a member variable. If it fails, it should show a proper warning
+     * and return gracefully.
+     */
+    exclusive = true;
+    return;
+}
+
+void ProfileIP::release_exclusive_ip_access(xclDeviceHandle handle, int index) {
+    /**
+     * TODO: when the XRT implements the exclusive context hal API, this
+     * method should close the previously requested exclusive context if
+     * one was request and then set the exclusive flag back to false. If
+     * it fails, it should show a proper warning and return gracefully.
+     */
+    exclusive = false;
+    return;
+}
+
+void ProfileIP::map() {
+    /**
+     * TODO: so far we are asking the debug_ip_layout where the IP is. Once the XRT hal implements
+     * the function that maps the IP registers to user memory space, this method should be simplified
+     * to one function call to the hal API and saves the result in a mapped_address and set mapped flag.
+     */
+    if (!exclusive) {
+        return;
+    }
+    std::string subdev = "icap";
+    std::string entry = "debug_ip_layout";
+    size_t max_path_size = 256;
+    char raw_debug_ip_layout_path[max_path_size];

```

Comment:
It's always a good idea to initialize variables when declared. Consider adding ' = {0}'  In this example the variable is assigned to a string after being populated by xclGetSysfsPath, can you count on the content being terminated?

Reply:
xclGetSysfsPath does make sure within the range it is null terminated, but I added null initialization and null termination at the end to make sure.